src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/base/SalesforceBaseSourceConfig.java

```diff
@@ -0,0 +1,145 @@
+  private Integer offset;
+
+  protected SalesforceBaseSourceConfig(String referenceName,
+                                       String clientId,
+                                       String clientSecret,
+                                       String username,
+                                       String password,
+                                       String loginUrl,
+                                       String errorHandling,
+                                       @Nullable String datetimeFilter,
+                                       @Nullable Integer duration,
+                                       @Nullable Integer offset) {
+    super(referenceName, clientId, clientSecret, username, password, loginUrl, errorHandling);
+    this.datetimeFilter = datetimeFilter;
+    this.duration = duration;
+    this.offset = offset;
+  }
+
+  public int getDuration() {
+    return Objects.isNull(duration) ? SalesforceSourceConstants.DURATION_DEFAULT : duration;
+  }
+
+  public int getOffset() {
+    return Objects.isNull(offset) ? SalesforceSourceConstants.OFFSET_DEFAULT : offset;
+  }
+
+  @Nullable
+  public String getDatetimeFilter() {
+    return datetimeFilter;
+  }
+
+  protected void validateFilters() {
+    validateSObjectFilter(SalesforceSourceConstants.PROPERTY_DURATION, getDuration());
+    validateSObjectFilter(SalesforceSourceConstants.PROPERTY_OFFSET, getOffset());
+  }
+
+  /**
+   * Generates SOQL based on given sObject name metadata and filter properties.
+   * Includes only those sObject fields which are present in the schema.
+   * This allows to avoid pulling data from Salesforce for the fields which are not needed.
+   *
+   * @param sObjectName Salesforce object name
+   * @param schema CDAP schema
+   *
+   * @return SOQL generated based on sObject metadata and given filters
+   */
+
+  protected String getSObjectQuery(String sObjectName, Schema schema) {
+    try {
+      SObjectDescriptor sObjectDescriptor = SObjectDescriptor.fromName(sObjectName, getAuthenticatorCredentials());
+
+      List<String> sObjectFields = sObjectDescriptor.getFieldsNames();
+
+      List<String> fieldNames;
+      if (schema == null) {
+        fieldNames = sObjectFields;
+      } else {
+        fieldNames = sObjectFields.stream()
+          .filter(name -> schema.getField(name) != null)
+          .collect(Collectors.toList());
+
+        if (fieldNames.isEmpty()) {
+          throw new IllegalArgumentException(
+            String.format("None of the fields indicated in schema are present in sObject metadata."
+              + " Schema: '%s'. SObject fields: '%s'", schema, sObjectFields));
+        }
+      }
+
+      String sObjectQuery = SalesforceQueryUtil.createSObjectQuery(fieldNames, sObjectName,
+        getDuration(), getOffset(), getDatetimeFilter());
+      LOG.debug("Generated SObject query: '{}'", sObjectQuery);
+      return sObjectQuery;
+    } catch (ConnectionException e) {
+      throw new IllegalStateException(
```

Comment:
doesn't look like this is adding much value, this method should just throw the ConnectionException.